src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -0,0 +1,168 @@
+            IEventAggregator eventAggregator,
+            IBuilder builder,
+            CriticalError criticalError,
+            string errorQueue)
+        {
+            this.configuration = configuration;
+            this.receiveInfrastructure = receiveInfrastructure;
+            this.queueBindings = queueBindings;
+            this.mainPipelineExecutor = mainPipelineExecutor;
+            this.eventAggregator = eventAggregator;
+            this.builder = builder;
+            this.criticalError = criticalError;
+            this.errorQueue = errorQueue;
+        }
+
+        public async Task Initialize()
+        {
+            if (!IsEnabled)
+            {
+                return;
+            }
+
+            if (configuration.PurgeOnStartup)
+            {
+                Logger.Warn("All queues owned by the endpoint will be purged on startup.");
+            }
+
+            AddReceivers();
+
+            foreach (var receiver in receivers)
+            {
+                try
+                {
+                    await receiver.Init().ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Fatal($"Receiver {receiver.Id} failed to initialize.", ex);
+                    throw;
+                }
+            }
+        }
+
+        public void Start()
+        {
+            foreach (var receiver in receivers)
+            {
+                try
+                {
+                    receiver.Start();
+                }
+                catch (Exception ex)
+                {
+                    Logger.Fatal($"Receiver {receiver.Id} failed to start.", ex);
+                    throw;
+                }
+            }
+        }
+
+        public Task Stop()
+        {
+            var receiverStopTasks = receivers.Select(async receiver =>
+            {
+                Logger.DebugFormat("Stopping {0} receiver", receiver.Id);
+                await receiver.Stop().ConfigureAwait(false);
+                Logger.DebugFormat("Stopped {0} receiver", receiver.Id);
+            });
+
+            return Task.WhenAll(receiverStopTasks);
+        }
+
+        public Task CreateQueuesIfNecessary(string username)
+        {
+            if (!IsEnabled)
+            {
+                return TaskEx.CompletedTask;
+            }
+
+            var queueCreator = receiveInfrastructure.QueueCreatorFactory();
+
+            return queueCreator.CreateQueueIfNecessary(queueBindings, username);
+        }
+
+        public async Task PerformPreStartupChecks()
+        {
+            if (!IsEnabled)
+            {
+                return;
+            }
+
+            var result = await receiveInfrastructure.PreStartupCheck().ConfigureAwait(false);
+
+            if (!result.Succeeded)
+            {
+                throw new Exception($"Pre start-up check failed: {result.ErrorMessage}");
+            }
+        }
+
+        bool IsEnabled => configuration != null;
```

Comment:
While I understand how send-only is currently being handled by passing back a null configuration from the `ReceiveConfigurationBuilder`, I'm not sure I like the how it means we're creating a dummy receive component even when we don't "need" receiving capabilities. It also leads to the need to handle a null config everywhere.

Feels like we're hiding it away when maybe we shouldn't be. 🤔

Reply:
Thinking about it a bit more, I might even be more OK with it as it currently is if `IsEnabled` was named `IsSendOnly` instead. Right now there's no concept of send-only in `ReceiveComponent` at all. If we're going to always create a `ReceiveComponent`, then I think it needs it more explicitly associated with why we're bailing out of some methods.

Reply:
@andreasohlund and I talked quite a while about this yesterday. While I'd rather see no receive component when the endpoint is send-only, this is super hard to figure out as an API consumer (even for internals) as the most components are always present while some might be null. I'd rather see some Maybe<ReceivingComponent> type of API to make this more clear but for now I think this approach is better than to expect everyone to do the null checks at the right place.